File: Algo/Leetcode/233NumOfOne.cpp
class Solution {
public:
    int countDigitOne(int n) {
        int count = 0;
        for(long long i=1; i<=n; i*=10){
            long long div = i * 10;
            count += (n / div) * i + min(max(n % div - i + 1, 0LL), i);
        }
        return count;
    }
};

// brute force
class Solution {
public:
    int countDigitOne(int n) {
        int count = 0;
        for(int i=1; i<=n; i++)
            count += countOne(i);
        return count;
    }

    int countOne(int i){
        string s = to_string(i);
        int count = 0;
        for(auto c : s) count+= (c=='1');
        return count;
    }
};

//
class Solution {
public:
    int countDigitOne(int n) {
        int count = 0;
        int a, b = 0;
        for(long m=1; m<=n; m*=10){
            a = n/m;
            b = n%m;
            count += (a+8)/10*m + (a%10==1)*(b+1);
            // count is for the 1 at each position.
            // (a+8)/10*m is by counting the before number from 0.
            // (a%10==1)*(b+1) is for counting the smaller fraction
        }
        return count;
    }
};

File: Algo/Leetcode/233NumOfOne.py
# Just brute force
class Solution(object):
    def countDigitOne(self, n):
        """
        :type n: int
        :rtype: int
        """
        count = 0
        for i in range(n+1):
            count += calOne(i)

        return count

def calOne(k):
    c = 0
    for i in str(k):
        if i == "1":
            c += 1

    return c

# each digit
class Solution(object):
    def countDigitOne(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n < 1:
            return 0

        s = str(n)
        return self.calOne(s)

    def calOne(self, s):
        n = len(s)
        if n == 1:
            return 0 if s == '0' else 1
        else:
            if s[0] == "0":  # no one on left most
                return self.calOne(s[1:])
            elif s[0] == "1":  # left most
                return self.calOne(s[1:]) + int(s[1:]) + 1 + 10 ** (n - 2) * (n - 1)
            else:
                return self.calOne(s[1:]) + 10 ** (n - 1) + 10 ** (n - 2) * (n - 1) * int(s[0])